CORE-10605 schema_registry: Implement enterprise request authz

pgellert opened the pull request:
This implements the authorization enforment logic inside the schema registry. Authorization can be enabled through the `schema_registry_enable_authorization` cluster config, in which case the schema registry starts respecting the configured ACLs on its endpoints. By default (when schema_registry_enable_authorization is off), the old approach of only checking for public/user-level/superuser-level access is performed.

For most endpoints, the authorization check is performed before entering the handler. For the two more complex handlers (get_subjects, get_schemas_ids_id) the check is deferred to only be done inside the handler, since for these endpoints, we prefer to adjust the response based on the ACLs the user has instead of outright reject the request.

Implementing the deferred authorization logic as well as integrating the ACL-based authorization is going to be implemented in follow up PRs (CORE-12275, CORE-12276, CORE-12277) to reduce the scope of this one.

Fixes: https://redpandadata.atlassian.net/browse/CORE-10605
Ref: https://github.com/redpanda-data/redpanda/pull/26446

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

BenPope committed:
schema_registry: Refactor: Extract auth from server

No functional change.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
schema_registry/auth: Refactor: Extract handle_authz

No functional change.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
config: Introduce schema_registry_enable_authorization

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
schema_registry/auth: Introduce enterprise authz

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
schema_registry/auth: Enforce enterprise auth

Implements: CORE-10605

Signed-off-by: Ben Pope <ben@redpanda.com>

pgellert committed:
dt/schema_registry: test ACL enforcement

Test the behaviour of ACL enforcement on the SR API.

michael-redpanda commented on src/v/pandaproxy/schema_registry/auth.h:
```diff
@@ -28,12 +33,53 @@ struct auth {
         superuser
     };
 
+    using global = global_resource;
```
question: will this need changing when your other PR lands?

pgellert replied:
Yes, I've now rebased and renamed this.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--9ce8beb9-d324-4fe0-827b-06a83558fc38-->
<!--TEST RESULTS-->
<!--2df1bd6e-513c-4951-a0c9-9b6a27276d30-->
<!--TEST RESULTS-->
<!--f53a77ff-efff-47d4-9461-d61390a899eb-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67907">build#67907</a></summary>


| test_class                    | test_method                               | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------------------|:------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterMetadataUploadTest     | test_uploads_after_restart                | {"cloud_storage_type": 2}                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67907#0197a889-edcc-4445-b683-dd0b652a465f | FLAKY         | 15/21    | upstream reliability is '93.66515837104072'. current run reliability is '71.42857142857143'. drift is 22.23659 and the allowed drift is set to 50. The test should PASS |
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67907#0197a889-edca-4376-b237-edb6cbcb9bdf | FLAKY         | 14/21    | upstream reliability is '69.98784933171325'. current run reliability is '66.66666666666666'. drift is 3.32118 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest      | test_node_operations                      | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67907#0197a889-edcc-4445-b683-dd0b652a465f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68005">build#68005</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68005#0197ad91-9d58-4149-89a3-2dda42604aef | FLAKY         | 20/21    | upstream reliability is '99.0909090909091'. current run reliability is '95.23809523809523'. drift is 3.85281 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68085">build#68085</a></summary>


| test_class                   | test_method                             | test_arguments                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:----------------------------------------|:-------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves | {"cleanup_policy": "compact,delete"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68085#0197b26e-db3b-4314-a94e-812be67dc5b4 | FLAKY         | 20/21    | upstream reliability is '97.70992366412213'. current run reliability is '95.23809523809523'. drift is 2.47183 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented on src/v/pandaproxy/schema_registry/auth.cc:
```diff
@@ -0,0 +1,214 @@
+security::audit::authentication_event_options make_authn_event_error(
+  const server::request_t& rq,
+  const ss::sstring& username,
+  ss::sstring reason) {
```
why is one value and one is reference?

pgellert replied:
No reason, good spot. I've now added a small commit to refactor this after the commit that moves this code: https://github.com/redpanda-data/redpanda/pull/26584/commits/87f7707a965a521b2738c8f54129e8fd912f450c

michael-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -3672,6 +3672,17 @@ configuration::configuration()
+      "Enable ACL-based authorization for Schema Registry requests. When true, "
+      "uses ACL-based authorization instead of the default "
+      "public/user/superuser authorization model. When false, uses the default "
+      "authorization model. Requires authentication to be enabled via "
+      "schema_registry_api.authn_method.",
```
> "Request authentication to be enabled via schema_registry_api.authn_method"

What does this mean?

michael-redpanda replied:
N/m I got authz/authn crossed in my head... again

michael-redpanda reviewed (commented):
this looks really good... just a couple of questions

pgellert committed:
schema_registry/auth: Refactor: string arg passing

Use std::string_view consistently instead of using a mix of passing by
value and by reference (and eventually still copying the string either
way).

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/7874943e5acc0d749bf5ee9c4ec71ac7d1378ae8..ff92138f5267b0dbab8d3db3fb6d48d08799073f):
* rebase to dev (to pull in https://github.com/redpanda-data/redpanda/pull/26585)
* address review comments above
* fix clang tidy error (use after move in auth.cc in constructor)

dotnwat commented on src/v/pandaproxy/schema_registry/auth.cc:
```diff
@@ -0,0 +1,214 @@
+
+using server = pandaproxy::ctx_server<pandaproxy::schema_registry::service>;
+
+inline net::unresolved_address from_ss_sa(const ss::socket_address& sa) {
+    return {fmt::format("{}", sa.addr()), sa.port(), sa.addr().in_family()};
+}
```
move to `net::`?

pgellert replied:
Done

dotnwat commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -6089,3 +6093,458 @@ def test_acl_idempotency(self):
+                request_fn = lambda auth: self._delete_subject(subject,
+                                                               auth=auth)
+                acl = self._create_acl(subject, "SUBJECT", "LITERAL", "REMOVE")
+                return request_fn, acl
```
nit: kind of looks like a lot of this could be presented by some sort of
```python
class Endpoint:
    name = "DELETE_SUBJECT"
    def make_request(self): ...
    def create_acl(self): ...
```

given that

```
        # No ACL — should be denied
        result = request_fn(self.user_auth)
        self.assert_equal(result.status_code, 403)

        # Grant correct ACL
        self._post_acl(acl)

        # Try again — should now succeed
        result = request_fn(self.user_auth)
        self.assert_equal(result.status_code, 200)
```

exists, suggests there is a common interface.

pgellert replied:
Thanks, this does look pretty nice. I've reworked this now

IoannisRP commented on src/v/pandaproxy/schema_registry/auth.cc:
```diff
@@ -0,0 +1,214 @@
+      rq.service().config().schema_registry_api.value(),
+      rq.req->get_listener_idx());
+
+    if (rq.authn_method != config::rest_authn_method::none) {
```
(pre-existing) nit

We can shave an indentation here
```
    if (rq.authn_method == config::rest_authn_method::none) {
        rq.user = credential_t{};
        audit_authn_success(rq);
        audit_authz_success(rq);
        return std::nullopt;
    }
    // rest of code...
```

pgellert replied:
I left this as is because the difference seems minor and I wanted to avoid the rebasing/fixup headache

IoannisRP commented on src/v/pandaproxy/schema_registry/auth.cc:
```diff
@@ -0,0 +1,214 @@
+        // did not give the authorization header.
+        if (config::shard_local_cfg().schema_registry_enable_authorization) {
+            if (is_deferred()) {
+                // The presence of a returned auth result indicates that AuthZ
+                // was deferred and has to be done inside the handler
+                return auth_result;
```
smell:
this is the only non-`nullopt` return, right? And it is deeply nested in the function.
It took me a while to understand that `nullopt` return means "it's all fine", the error handling is done through exceptions and the optional is only for the deferred case.

If it's not easy to rework this logic into something more palatable, i think a comment explaining the above next to the definition would be very helpful..

pgellert replied:
I've added a comment now to the header to try to hopefully make the behaviour clearer now

IoannisRP commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -46,39 +47,47 @@
+      , _h{std::move(h)} {
+        const auto is_h_deferred
+          = std::holds_alternative<auth::deferred_function_handler>(_h);
+        vassert(
+          (_auth.is_deferred() && is_h_deferred)
+            || (!_auth.is_deferred() && !is_h_deferred),
+          "Deferred auth endpoints must use a deferred handler");
```
```suggestion
        vassert(
          _auth.is_deferred() == is_h_deferred,
          "Deferred auth endpoints must use a deferred handler");
```

IoannisRP reviewed (commented):
Just a few nits

pgellert committed:
net: turn from_ss_sa into a constructor

Refactor for easier reuse.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/ff92138f5267b0dbab8d3db3fb6d48d08799073f..27749d61e5d54f5a976a07a52577a5cb62caef39): address feedback
